typed/union_test.go

```diff
@@ -0,0 +1,224 @@
+		},
+		{
+			name: "proper union update from not-set, setting discriminator",
+			old:  `{}`,
+			new:  `{"b": 1}`,
+			out:  `{"b": 1, "discriminator": "b"}`,
+		},
+		{
+			name: "proper union update from not-set, no discriminator",
+			old:  `{}`,
+			new:  `{"two": 1}`,
+			out:  `{"two": 1}`,
+		},
+		{
+			name: "remove union, with discriminator",
+			old:  `{"a": 1}`,
+			new:  `{}`,
+			out:  `{}`,
+		},
+		{
+			name: "remove union and discriminator",
+			old:  `{"a": 1, "discriminator": "a"}`,
+			new:  `{}`,
+			out:  `{}`,
+		},
+		{
+			name: "remove union, not discriminator",
+			old:  `{"a": 1, "discriminator": "a"}`,
+			new:  `{"discriminator": "a"}`,
+			out:  `{"discriminator": "a"}`,
+		},
+		{
+			name: "remove union, no discriminator",
+			old:  `{"two": 1}`,
+			new:  `{}`,
+			out:  `{}`,
+		},
+		{
+			name: "dumb client update, no discriminator",
+			old:  `{"one": 1}`,
+			new:  `{"one": 2, "two": 1}`,
+			out:  `{"two": 1}`,
+		},
+		{
+			name: "dumb client update, sets discriminator",
+			old:  `{"a": 1}`,
+			new:  `{"a": 2, "b": 1}`,
+			out:  `{"b": 1, "discriminator": "b"}`,
+		},
+		{
+			name: "dumb client doesn't update discriminator",
+			old:  `{"a": 1, "discriminator": "a"}`,
+			new:  `{"a": 2, "b": 1, "discriminator": "a"}`,
+			out:  `{"b": 1, "discriminator": "b"}`,
+		},
+		{
+			name: "client sends multiple with discriminator",
+			old:  `{}`,
+			new:  `{"a": 1, "b": 1, "discriminator": "a"}`,
+			out:  `{"a": 1, "discriminator": "a"}`,
+		},
+		{
+			name: "multi-discriminator at the same time",
+			old:  `{"one": 1, "a": 1}`,
+			new:  `{"one": 1, "two": 1, "a": 1, "b": 1}`,
+			out:  `{"two": 1, "b": 1, "discriminator": "b"}`,
+		},
+		{
+			name: "change discriminator, nothing else",
+			old:  `{"discriminator": "a"}`,
+			new:  `{"discriminator": "random"}`,
+			out:  `{"discriminator": "random"}`,
+		},
+		{
+			name: "change discriminator, nothing else, it drops other field",
+			old:  `{"discriminator": "a", "a": 1}`,
+			new:  `{"discriminator": "random", "a": 1}`,
+			out:  `{"discriminator": "random"}`,
+		},
+		{
+			name: "remove discriminator, nothing else",
+			old:  `{"discriminator": "a", "a": 1}`,
+			new:  `{"a": 1}`,
+			out:  `{"a": 1}`,
+		},
+		{
+			name: "remove discriminator, add new field",
+			old:  `{"discriminator": "a", "a": 1}`,
+			new:  `{"b": 1}`,
+			out:  `{"b": 1, "discriminator": "b"}`,
+		},
+		{
+			name: "old object has two of same union, but we add third",
+			old:  `{"discriminator": "a", "a": 1, "b": 1}`,
+			new:  `{"discriminator": "a", "a": 1, "b": 1, "c": 1}`,
+			out:  `{"discriminator": "c", "c": 1}`,
+		},
+		// Following use-cases are not doing anything because they are "broken"
+		{
+			name: "old object has two of same union set",
+			old:  `{"a": 1, "b": 2}`,
+			new:  `{"a": 2, "b": 1}`,
+			out:  `{"a": 2, "b": 1}`,
+		},
+		{
+			name: "new object has three of same union set",
+			old:  `{"a": 1}`,
+			new:  `{"a": 2, "b": 1, "c": 3}`,
```

Comment:
This should cause an error, I think.

Reply:
I think this would cause an error outside smd as well